flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/asyncprocessing/AbstractAsyncStateStreamOperator.java

```diff
@@ -220,6 +222,40 @@ public <K, N> InternalTimerService<N> getInternalTimerService(
                 (AsyncExecutionController<K>) asyncExecutionController);
     }
 
+    @Override
+    public void processWatermark(Watermark mark) throws Exception {
+        if (!isAsyncStateProcessingEnabled()) {
+            // If async state processing is disabled, fallback to the super class.
+            super.processWatermark(mark);
+            return;
+        }
+        asyncExecutionController
+                .getEpochManager()
+                .onNonRecord(
+                        () -> {
+                            try {
+                                if (timeServiceManager != null) {
+                                    timeServiceManager.advanceWatermark(mark);
+                                }
+                                output.emitWatermark(mark);
+                            } catch (Exception e) {
+                                // todo: error handling
+                            }
+                        });
+    }
+
+    @Override
+    public void processWatermarkStatus(WatermarkStatus watermarkStatus) throws Exception {
```

Comment:
Apart from Watermark and WatermarkStatus, there are other events that could come through the data stream(See all subclasses of StreamElement and RuntimeEvent). This PR might need to handle all these events, and it needs some design how to treat all events in a unified way.

Reply:
`InternalWatermark` also uses `AbstractStreamOperator#processWatermark`.
For `LatencyMarker` `EndOfDataMarker` and `RecordAttributes`, I plan to do it in FLINK-35031.